dynawo/sources/Models/CPP/Controls/Shunts/ModelCentralizedShuntsSectionControl/DYNModelCentralizedShuntsSectionControl.cpp

```diff
@@ -0,0 +1,346 @@
+// for power systems.
+//
+
+/**
+ * @file  DYNModelCentralizedShuntsSectionControl.cpp
+ *
+ * @brief centralized model for shunt section control implementation
+ *
+ */
+
+#include <sstream>
+#include <vector>
+#include <algorithm>
+
+#include "PARParametersSet.h"
+
+#include "DYNNumericalUtils.h"
+#include "DYNModelCentralizedShuntsSectionControl.h"
+#include "DYNModelCentralizedShuntsSectionControl.hpp"
+#include "DYNSparseMatrix.h"
+#include "DYNMacrosMessage.h"
+#include "DYNElement.h"
+#include "DYNCommonModeler.h"
+#include "DYNTrace.h"
+#include "DYNVariableForModel.h"
+#include "DYNParameter.h"
+
+using std::vector;
+using std::string;
+using std::map;
+using std::stringstream;
+
+using boost::shared_ptr;
+
+using parameters::ParametersSet;
+
+extern "C" DYN::SubModelFactory* getFactory() {
+  return (new DYN::ModelCentralizedShuntsSectionControlFactory());
+}
+
+extern "C" void deleteFactory(DYN::SubModelFactory* factory) {
+  delete factory;
+}
+
+extern "C" DYN::SubModel* DYN::ModelCentralizedShuntsSectionControlFactory::create() const {
+  DYN::SubModel* model(new DYN::ModelCentralizedShuntsSectionControl());
+  return model;
+}
+
+extern "C" void DYN::ModelCentralizedShuntsSectionControlFactory::destroy(DYN::SubModel* model) const {
+  delete model;
+}
+
+namespace DYN {
+  ModelCentralizedShuntsSectionControl::ModelCentralizedShuntsSectionControl() : Impl("CentralizedShuntsSectionControl"),
+    nbShunts_(0),
+    isSelf_(false),
+    URef0Pu_(0.),
+    changingShunt(-1),
+    whenUp_(VALDEF),
+    whenDown_(VALDEF),
+    tNext_(10.),
+    lastTime_(VALDEF) {}
+
+  void
+  ModelCentralizedShuntsSectionControl::defineParameters(std::vector<ParameterModeler>& parameters) {
+    parameters.push_back(ParameterModeler("nbShunts", VAR_TYPE_INT, EXTERNAL_PARAMETER));
+    parameters.push_back(ParameterModeler("section0", VAR_TYPE_INT, EXTERNAL_PARAMETER, "*", "nbShunts"));
+    parameters.push_back(ParameterModeler("SectionMin", VAR_TYPE_INT, EXTERNAL_PARAMETER, "*", "nbShunts"));
+    parameters.push_back(ParameterModeler("SectionMax", VAR_TYPE_INT, EXTERNAL_PARAMETER, "*", "nbShunts"));
+    parameters.push_back(ParameterModeler("DeadbandUPu", VAR_TYPE_DOUBLE, EXTERNAL_PARAMETER, "*", "nbShunts"));
+    parameters.push_back(ParameterModeler("IsSelf", VAR_TYPE_BOOL, EXTERNAL_PARAMETER));
+    parameters.push_back(ParameterModeler("URef0Pu", VAR_TYPE_DOUBLE, EXTERNAL_PARAMETER));
+    parameters.push_back(ParameterModeler("tNext", VAR_TYPE_DOUBLE, EXTERNAL_PARAMETER));
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::setSubModelParameters() {
+    try {
+      nbShunts_ = findParameterDynamic("nbShunts").getValue<int>();
+      isSelf_ = findParameterDynamic("IsSelf").getValue<bool>();
+      URef0Pu_ = findParameterDynamic("URef0Pu").getValue<double>();
+      tNext_ = findParameterDynamic("tNext").getValue<double>();
+      for (int s = 0; s < nbShunts_; ++s) {
+        std::stringstream section0Name;
```

Comment:
Put stringstream outside the for loop and use `section0Name.str(std::string());` and `section0Name.clear();`.

I have tested it here : https://quick-bench.com/q/8yGzFheZCCc0G19fFnMeDh1uu38